dotnet/src/SemanticKernel/Functions/FunctionCollection.cs

```diff
@@ -0,0 +1,123 @@
+    {
+        this._logger = loggerFactory is not null ? loggerFactory.CreateLogger(typeof(FunctionCollection)) : NullLogger.Instance;
+
+        // Important: names are case insensitive
+        this._functionCollection = new(StringComparer.OrdinalIgnoreCase);
+    }
+
+    /// <summary>
+    /// Adds a function to the function collection.
+    /// </summary>
+    /// <param name="functionInstance">The function instance to add.</param>
+    /// <returns>The updated function collection.</returns>
+    public IFunctionCollection AddFunction(ISKFunction functionInstance)
+    {
+        Verify.NotNull(functionInstance);
+
+        ConcurrentDictionary<string, ISKFunction> functions = this._functionCollection.GetOrAdd(functionInstance.PluginName, static _ => new(StringComparer.OrdinalIgnoreCase));
+        functions[functionInstance.Name] = functionInstance;
+
+        return this;
+    }
+
+    /// <inheritdoc/>
+    public ISKFunction GetFunction(string functionName) =>
+        this.GetFunction(GlobalFunctions, functionName);
+
+    /// <inheritdoc/>
+    public ISKFunction GetFunction(string pluginName, string functionName)
+    {
+        if (!this.TryGetFunction(pluginName, functionName, out ISKFunction? functionInstance))
+        {
+            this.ThrowFunctionNotAvailable(pluginName, functionName);
+        }
+
+        return functionInstance;
+    }
+
+    /// <inheritdoc/>
+    public bool TryGetFunction(string functionName, [NotNullWhen(true)] out ISKFunction? availableFunction) =>
+        this.TryGetFunction(GlobalFunctions, functionName, out availableFunction);
+
+    /// <inheritdoc/>
+    public bool TryGetFunction(string pluginName, string functionName, [NotNullWhen(true)] out ISKFunction? availableFunction)
+    {
+        Verify.NotNull(pluginName);
+        Verify.NotNull(functionName);
+
+        if (this._functionCollection.TryGetValue(pluginName, out ConcurrentDictionary<string, ISKFunction>? functions))
+        {
+            return functions.TryGetValue(functionName, out availableFunction);
+        }
+
+        availableFunction = null;
+        return false;
+    }
+
+    /// <inheritdoc/>
+    public IReadOnlyList<FunctionView> GetFunctionViews()
+    {
+        var result = new List<FunctionView>();
+
+        foreach (var functions in this._functionCollection.Values)
+        {
+            foreach (ISKFunction f in functions.Values)
+            {
+                result.Add(f.Describe());
+            }
+        }
+
+        return result;
+    }
+
+    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+    internal string DebuggerDisplay => $"Count = {this._functionCollection.Count}";
+
+    #region private ================================================================================
+
+    [DoesNotReturn]
+    private void ThrowFunctionNotAvailable(string pluginName, string functionName)
+    {
+        this._logger.LogError("Function not available: plugin:{0} function:{1}", pluginName, functionName);
```

Comment:
Change to template on affected code.

```suggestion
        this._logger.LogError("Function not available: plugin:{PluginName} function:{FunctionName}", pluginName, functionName);
```